File: wp-content/prerf/analyze-har-files.php
<?php
/**
 * HAR File Analysis Script
 * Analyzes HAR files for performance metrics
 */

// HAR files to analyze
$harFiles = [
    'home' => 'trang-chu.har',
    'products' => 'trang-san-pham.har', 
    'post' => 'trang-bai-viet.har',
    'contact' => 'trang-contact.har',
    'cart' => 'trang-cart.har',
    'checkout' => 'trang-checkout.har'
];

$results = [];

foreach ($harFiles as $route => $filename) {
    $filePath = __DIR__ . '/inputs/' . $filename;
    
    if (!file_exists($filePath)) {
        echo "HAR file not found: $filePath\n";
        continue;
    }
    
    $harContent = file_get_contents($filePath);
    $harData = json_decode($harContent, true);
    
    if (!$harData || !isset($harData['log']['entries'])) {
        echo "Invalid HAR format for: $filename\n";
        continue;
    }
    
    $entries = $harData['log']['entries'];
    $totalRequests = count($entries);
    
    // Initialize counters
    $errors404 = [];
    $recaptchaCount = 0;
    $cssCount = 0;
    $jsCount = 0;
    $duplicates = [];
    $urlCounts = [];
    
    // Analyze each request
    foreach ($entries as $entry) {
        $url = $entry['request']['url'];
        $status = $entry['response']['status'];
        $mimeType = $entry['response']['content']['mimeType'] ?? '';
        
        // Count URLs for duplicate detection
        $urlCounts[$url] = ($urlCounts[$url] ?? 0) + 1;
        
        // Check for 404 errors
        if ($status == 404) {
            $errors404[] = parse_url($url, PHP_URL_PATH) ?: $url;
        }
        
        // Check for reCAPTCHA and gstatic
        if (stripos($url, 'recaptcha') !== false || stripos($url, 'gstatic.com') !== false) {
            $recaptchaCount++;
        }
        
        // Count CSS files
        if (stripos($mimeType, 'css') !== false || 
            preg_match('/\.css(\?|$)/', $url)) {
            $cssCount++;
        }
        
        // Count JS files
        if (stripos($mimeType, 'javascript') !== false || 
            stripos($mimeType, 'ecmascript') !== false ||
            preg_match('/\.js(\?|$)/', $url)) {
            $jsCount++;
        }
    }
    
    // Find duplicates
    foreach ($urlCounts as $url => $count) {
        if ($count > 1) {
            $duplicates[$url] = $count;
        }
    }
    
    $results[$route] = [
        'total_requests' => $totalRequests,
        '404_errors' => [
            'count' => count($errors404),
            'files' => $errors404
        ],
        'recaptcha_loads' => $recaptchaCount,
        'css_files' => $cssCount,
        'js_files' => $jsCount,
        'duplicate_requests' => [
            'count' => count($duplicates),
            'details' => $duplicates
        ]
    ];
}

// Previous V2 baseline data
$previousV2 = [
    '404_errors' => '2-4',
    'recaptcha_loads' => '14-27',
    'cart_requests' => 251,
    'checkout_requests' => 242
];

// Target metrics
$targets = [
    '404_errors' => 0,
    'recaptcha_loads' => 1,
    'cart_requests' => '<150',
    'checkout_requests' => '<180'
];

// Display results
echo "\n=== HAR FILE ANALYSIS RESULTS ===\n\n";

foreach ($results as $route => $metrics) {
    echo "--- $route ---\n";
    echo "Total Requests: {$metrics['total_requests']}\n";
    echo "404 Errors: {$metrics['404_errors']['count']}";
    if (!empty($metrics['404_errors']['files'])) {
        echo " (Files: " . implode(', ', array_slice($metrics['404_errors']['files'], 0, 3));
        if (count($metrics['404_errors']['files']) > 3) {
            echo ", +" . (count($metrics['404_errors']['files']) - 3) . " more";
        }
        echo ")";
    }
    echo "\n";
    echo "reCAPTCHA Loads: {$metrics['recaptcha_loads']}\n";
    echo "CSS Files: {$metrics['css_files']}\n";
    echo "JS Files: {$metrics['js_files']}\n";
    echo "Duplicate Requests: {$metrics['duplicate_requests']['count']}\n";
    echo "\n";
}

// Comparison Summary
echo "=== COMPARISON WITH TARGETS ===\n\n";
echo "Previous V2 Baseline:\n";
echo "- 404 Errors: 2-4\n";
echo "- reCAPTCHA: 14-27\n";
echo "- Cart: 251 requests\n";
echo "- Checkout: 242 requests\n\n";

echo "Current Results:\n";
$all404s = 0;
$allRecaptcha = 0;
foreach ($results as $route => $metrics) {
    $all404s += $metrics['404_errors']['count'];
    $allRecaptcha = max($allRecaptcha, $metrics['recaptcha_loads']);
}
echo "- 404 Errors: $all404s (Target: 0) " . ($all404s == 0 ? "✓ MET" : "✗ NOT MET") . "\n";
echo "- reCAPTCHA: $allRecaptcha (Target: 1) " . ($allRecaptcha <= 1 ? "✓ MET" : "✗ NOT MET") . "\n";

if (isset($results['cart'])) {
    $cartTotal = $results['cart']['total_requests'];
    echo "- Cart: $cartTotal requests (Target: <150) " . ($cartTotal < 150 ? "✓ MET" : "✗ NOT MET") . "\n";
}

if (isset($results['checkout'])) {
    $checkoutTotal = $results['checkout']['total_requests'];
    echo "- Checkout: $checkoutTotal requests (Target: <180) " . ($checkoutTotal < 180 ? "✓ MET" : "✗ NOT MET") . "\n";
}

// Save results to JSON
$outputData = [
    'analysis_date' => date('Y-m-d H:i:s'),
    'results' => $results,
    'comparison' => [
        'previous_v2' => $previousV2,
        'targets' => $targets,
        'target_met' => [
            '404_errors' => $all404s == 0,
            'recaptcha_loads' => $allRecaptcha <= 1,
            'cart_requests' => isset($results['cart']) ? $results['cart']['total_requests'] < 150 : null,
            'checkout_requests' => isset($results['checkout']) ? $results['checkout']['total_requests'] < 180 : null
        ]
    ]
];

file_put_contents(__DIR__ . '/har-analysis-output.json', json_encode($outputData, JSON_PRETTY_PRINT));
echo "\nDetailed results saved to: har-analysis-output.json\n";

File: wp-content/prerf/analyze-har-files.py
#!/usr/bin/env python3
"""
HAR File Analysis Script
Analyzes HAR files for performance metrics
"""

import json
import os
from collections import defaultdict
from urllib.parse import urlparse

# HAR files to analyze
har_files = {
    'home': 'trang-chu.har',
    'products': 'trang-san-pham.har',
    'post': 'trang-bai-viet.har',
    'contact': 'trang-contact.har',
    'cart': 'trang-cart.har',
    'checkout': 'trang-checkout.har'
}

results = {}
base_dir = os.path.dirname(os.path.abspath(__file__))
inputs_dir = os.path.join(base_dir, 'inputs')

for route, filename in har_files.items():
    file_path = os.path.join(inputs_dir, filename)
    
    if not os.path.exists(file_path):
        print(f"HAR file not found: {file_path}")
        continue
    
    with open(file_path, 'r', encoding='utf-8') as f:
        try:
            har_data = json.load(f)
        except json.JSONDecodeError as e:
            print(f"Invalid JSON in {filename}: {e}")
            continue
    
    if 'log' not in har_data or 'entries' not in har_data['log']:
        print(f"Invalid HAR format for: {filename}")
        continue
    
    entries = har_data['log']['entries']
    total_requests = len(entries)
    
    # Initialize counters
    errors_404 = []
    recaptcha_count = 0
    css_count = 0
    js_count = 0
    url_counts = defaultdict(int)
    
    # Analyze each request
    for entry in entries:
        url = entry['request']['url']
        status = entry['response']['status']
        mime_type = entry['response']['content'].get('mimeType', '')
        
        # Count URLs for duplicate detection
        url_counts[url] += 1
        
        # Check for 404 errors
        if status == 404:
            parsed_url = urlparse(url)
            errors_404.append(parsed_url.path if parsed_url.path else url)
        
        # Check for reCAPTCHA and gstatic
        url_lower = url.lower()
        if 'recaptcha' in url_lower or 'gstatic.com' in url_lower:
            recaptcha_count += 1
        
        # Count CSS files
        if 'css' in mime_type.lower() or '.css' in url:
            css_count += 1
        
        # Count JS files
        mime_lower = mime_type.lower()
        if 'javascript' in mime_lower or 'ecmascript' in mime_lower or '.js' in url:
            js_count += 1
    
    # Find duplicates
    duplicates = {url: count for url, count in url_counts.items() if count > 1}
    
    results[route] = {
        'total_requests': total_requests,
        '404_errors': {
            'count': len(errors_404),
            'files': errors_404
        },
        'recaptcha_loads': recaptcha_count,
        'css_files': css_count,
        'js_files': js_count,
        'duplicate_requests': {
            'count': len(duplicates),
            'details': duplicates
        }
    }

# Display results
print("\n=== HAR FILE ANALYSIS RESULTS ===\n")

for route, metrics in results.items():
    print(f"--- {route} ---")
    print(f"Total Requests: {metrics['total_requests']}")
    print(f"404 Errors: {metrics['404_errors']['count']}", end='')
    if metrics['404_errors']['files']:
        files_preview = metrics['404_errors']['files'][:3]
        print(f" (Files: {', '.join(files_preview)}", end='')
        if len(metrics['404_errors']['files']) > 3:
            print(f", +{len(metrics['404_errors']['files']) - 3} more", end='')
        print(")", end='')
    print()
    print(f"reCAPTCHA Loads: {metrics['recaptcha_loads']}")
    print(f"CSS Files: {metrics['css_files']}")
    print(f"JS Files: {metrics['js_files']}")
    print(f"Duplicate Requests: {metrics['duplicate_requests']['count']}")
    print()

# Comparison Summary
print("=== COMPARISON WITH TARGETS ===\n")
print("Previous V2 Baseline:")
print("- 404 Errors: 2-4")
print("- reCAPTCHA: 14-27")
print("- Cart: 251 requests")
print("- Checkout: 242 requests\n")

print("Current Results:")
all_404s = sum(metrics['404_errors']['count'] for metrics in results.values())
all_recaptcha = max(metrics['recaptcha_loads'] for metrics in results.values())

print(f"- 404 Errors: {all_404s} (Target: 0) {'✓ MET' if all_404s == 0 else '✗ NOT MET'}")
print(f"- reCAPTCHA: {all_recaptcha} (Target: 1) {'✓ MET' if all_recaptcha <= 1 else '✗ NOT MET'}")

if 'cart' in results:
    cart_total = results['cart']['total_requests']
    print(f"- Cart: {cart_total} requests (Target: <150) {'✓ MET' if cart_total < 150 else '✗ NOT MET'}")

if 'checkout' in results:
    checkout_total = results['checkout']['total_requests']
    print(f"- Checkout: {checkout_total} requests (Target: <180) {'✓ MET' if checkout_total < 180 else '✗ NOT MET'}")

# Save results to JSON
output_data = {
    'analysis_date': str(__import__('datetime').datetime.now()),
    'results': results,
    'comparison': {
        'previous_v2': {
            '404_errors': '2-4',
            'recaptcha_loads': '14-27',
            'cart_requests': 251,
            'checkout_requests': 242
        },
        'targets': {
            '404_errors': 0,
            'recaptcha_loads': 1,
            'cart_requests': '<150',
            'checkout_requests': '<180'
        },
        'target_met': {
            '404_errors': all_404s == 0,
            'recaptcha_loads': all_recaptcha <= 1,
            'cart_requests': results['cart']['total_requests'] < 150 if 'cart' in results else None,
            'checkout_requests': results['checkout']['total_requests'] < 180 if 'checkout' in results else None
        }
    }
}

with open(os.path.join(base_dir, 'har-analysis-output.json'), 'w') as f:
    json.dump(output_data, f, indent=2)

print("\nDetailed results saved to: har-analysis-output.json")